core/services/ping/pingutils.py

```diff
@@ -36,12 +39,26 @@ class PingDeviceDescriptor:
     firmware_version_minor: int
     firmware_version_patch: int
     port: SysFS
+    ethernet_discovery_info: Optional[str]
+    driver: Optional["PingDriver"]  # type: ignore
+
+    def get_hw_or_eth_info(self) -> str:
+        if self.port:
+            return str(self.port.device_path)
+        return str(self.ethernet_discovery_info)
 
     def __hash__(self) -> int:
-        return hash(self.port.hwid)
+        return hash(self.get_hw_or_eth_info())
 
     def __str__(self) -> str:
         return f"""{self.ping_type.name}
 ID: {self.device_id}
 FW: v{self.firmware_version_major}.{self.firmware_version_minor}.{self.firmware_version_patch}
-port: {self.port.hwid}"""
+port: {self.get_hw_or_eth_info()}"""
+
+
+def udp_port_is_in_use(port: int) -> bool:
+    for conn in psutil.net_connections():
+        if conn.laddr.port == port and conn.type == socket.SocketKind.SOCK_DGRAM:
+            return True
+    return False
```

Comment:
```suggestion
    return any(conn.laddr.port == port and conn.type == socket.SocketKind.SOCK_DGRAM
               for conn in psutil.net_connections())
```

Reply:
I always forget about any/some :man_shrugging:

Reply:
`any` and `all` are quite useful. Same with the boolean operators (`or`, `and`). Short-circuiting things are oddly satisfying to me :-)